src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs

```diff
@@ -415,6 +422,152 @@ private void HandleCodeActionRequest(string id, string paramsJson)
+                // No tokens for invalid range
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var uri = new Uri(semanticTokensParams.TextDocument.Uri);
+            var queryParams = HttpUtility.ParseQueryString(uri.Query);
+            var expression = queryParams.Get("expression");
+
+            if (string.IsNullOrEmpty(expression))
+            {
+                // Empty tokens for the empty expression
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            // Monaco-Editor sometimes uses \r\n for the newline character. \n is not always the eol character so allowing clients to pass eol character
+            var eol = queryParams.Get("eol");
+            eol = !string.IsNullOrEmpty(eol) ? eol : EOL.ToString();
+
+            var (startIndex, endIndex) = PositionRangeHelper.ConvertRangeToPositions(semanticTokensParams.Range, expression, eol);
+            if (startIndex < 0 || endIndex < 0)
+            {
+                SendEmptySemanticTokensResponse(id);
+                return;
+            }
+
+            var tokenTypesToSkip = ParseTokenTypesToSkipParam(queryParams.Get("tokenTypesToSkip"));
+            var scope = _scopeFactory.GetOrCreateInstance(semanticTokensParams.TextDocument.Uri);
+            var result = scope.Check(expression);
+
+            // Skip over the token types that clients don't want in the response
+            var tokens = result.GetTokens().Where(tok => !tokenTypesToSkip.Contains(tok.TokenType));
+
+            // Only consider overlapping tokens. end index is exlcusive
+            var overlappingTokens = tokens.Where(token => !(token.EndIndex <= startIndex || token.StartIndex >= endIndex));
+            var encodedTokens = SemanticTokensEncoder.EncodeTokens(overlappingTokens, expression, eol);
+
+            _sendToClient(JsonRpcHelper.CreateSuccessResult(id, new SemanticTokensResponse() { Data = encodedTokens }));
+        }
+
+        private static ICollection<TokenType> ParseTokenTypesToSkipParam(string rawTokenTypesToSkipParam)
+        {
+            var tokenTypesToSkip = new HashSet<TokenType>();    
+            if (string.IsNullOrWhiteSpace(rawTokenTypesToSkipParam))
+            {
+                return tokenTypesToSkip;
+            }
+
+            try
+            {
+                var tokenTypesToSkipParam = JsonSerializer.Deserialize<List<int>>(rawTokenTypesToSkipParam);
+                foreach (var tokenTypeValue in tokenTypesToSkipParam)
+                {
+                    var tokenType = (TokenType)tokenTypeValue;
+                    if (tokenType != TokenType.Lim)
+                    {
+                        tokenType = tokenType == TokenType.Min ? TokenType.Unknown : tokenType;
+                        tokenTypesToSkip.Add(tokenType);
+                    }
+                }
+
+                return tokenTypesToSkip;
+            }
+            catch
```

Comment:
>catch [](http://example.com/codeflow?start=12&length=5)

Should we be specific here about what exceptions we intend to catch?

Reply:
I debated that but we really don't need to catch it and log it as failure to parse tokenTypesToSkip param is considered as not skipping any tokenTypes. But i agree the general good practice is to be specific. I will change this

Reply:
Catching specific type of exceptions